packages/typescript-estree/src/source-files.ts

```diff
@@ -0,0 +1,14 @@
+import * as ts from 'typescript';
+
+export function isSourceFile(code: unknown): code is ts.SourceFile {
+  return (
+    typeof code !== 'string' &&
+    (code as Partial<ts.SourceFile> | undefined)?.kind ===
+      ts.SyntaxKind.SourceFile &&
+    !!(code as Partial<ts.SourceFile>).getFullText
+  );
+}
```

Comment:
I'd probably do something like this to avoid multiple casts:

```suggestion
export function isSourceFile(code: unknown): code is ts.SourceFile {
  if (typeof code !== 'object' || code == null) {
    return false;
  }
  
  const maybeSourceFile = code as Partial<ts.SourceFile>;
  return (
    maybeSourceFile.kind === ts.SyntaxKind.SourceFile &&
    typeof maybeSourceFile.getFullText === 'function'
  );
}
```

Reply:
🤔 My personal weights have making an extra variable for an `as` as slightly more undesirable than an extra `as`. Is this something you feel strongly about?

Reply:
IMO an additional variable is better because it reduces duplication which reduces chance for typos.

For me reading and parsing a long logical is not super simple.
Parsing it with two assertions nested within it is more difficult.

A new variable is pretty-well zero-cost at runtime, so it's not a problem to add one for the sake of clarity, IMO.